lib/discourse_slack/api.rb

```diff
@@ -0,0 +1,72 @@
+module DiscourseSlack
+  class API
+
+    BASE_URL = "https://slack.com/api/"
+
+    def self.http
+      @http ||= begin
+        http = Net::HTTP.new("slack.com" , 443)
+        http.use_ssl = true
+
+        http
+      end
+    end
+
+    def self.get_request(uri)
+      response = http.request(Net::HTTP::Get.new(uri))
+
+      return JSON(response.body) if response && response.code == "200"
+
+      { "error": I18n.t('slack.errors.invalid_response') }
+    end
+
+    def self.api_uri(method, params = {})
+      params.merge!(token_param)
+      URI("#{BASE_URL}#{method}?#{params.to_query}")
+    end
+
+    def self.channels
+      return [] if SiteSetting.slack_access_token.empty?
+
+      @channels = Rails.cache.fetch("slack_channels", expires_in: 15.minutes) do
```

Comment:
We should refresh the cache in a background job as well. We don't want to be waiting for the request to slack each time we burst the cache.